source/common/http/conn_manager_impl.cc

```diff
@@ -1135,6 +1135,73 @@ void ConnectionManagerImpl::ActiveStream::snapScopedRouteConfig() {
 
 void ConnectionManagerImpl::ActiveStream::refreshCachedRoute() { refreshCachedRoute(nullptr); }
 
+void ConnectionManagerImpl::ActiveStream::refreshDurationTimeout() {
+  if (!filter_manager_.streamInfo().route_entry_ || !request_headers_) {
+    return;
+  }
+  auto& route = filter_manager_.streamInfo().route_entry_;
+
+  auto grpc_timeout = Grpc::Common::getGrpcTimeout(*request_headers_);
+  std::chrono::milliseconds timeout = std::chrono::milliseconds(0);
+
+  if (!grpc_timeout || !route->grpcTimeoutHeaderMax()) {
+    // Either there is no grpc-timeout header or special timeouts for it are not
+    // configured. Use stream duration.
+    if (route->maxStreamDuration()) {
+      timeout = route->maxStreamDuration().value();
+    } else {
+      // Fall back to HCM config.
+      const auto max_stream_duration = connection_manager_.config_.maxStreamDuration();
+      if (max_stream_duration.has_value() && max_stream_duration.value().count()) {
+        timeout = max_stream_duration.value();
+      }
+    }
+  } else {
+    // Start with the timeout equal to the gRPC timeout header.
+    timeout = grpc_timeout.value();
+    // If there's a valid cap, apply it.
+    if ((timeout > route->grpcTimeoutHeaderMax().value() ||
+         timeout == std::chrono::milliseconds(0)) &&
+        route->grpcTimeoutHeaderMax().value() != std::chrono::milliseconds(0)) {
+      timeout = route->grpcTimeoutHeaderMax().value();
+    }
+
+    // Use the configured offset iff the timeout will stay positive.
+    if (timeout != std::chrono::milliseconds(0) && route->grpcTimeoutHeaderOffset()) {
+      const auto offset = route->grpcTimeoutHeaderOffset().value();
+      if (offset < timeout) {
+        timeout -= offset;
+      }
+    }
+  }
+  // If the timeout is infinite, unregister any existing max duration timer.
+  if (timeout == std::chrono::milliseconds(0)) {
+    if (max_stream_duration_timer_) {
+      max_stream_duration_timer_->disableTimer();
+    }
+    return;
+  }
+
+  // See how long this stream has been alive, and adjust the timeout
+  // accordingly.
+  std::chrono::duration time_used = std::chrono::duration_cast<std::chrono::milliseconds>(
+      connection_manager_.timeSource().monotonicTime() -
+      filter_manager_.streamInfo().startTimeMonotonic());
+  if (timeout > time_used) {
+    timeout -= time_used;
+  } else {
+    timeout = std::chrono::milliseconds(0);
+  }
+
+  // Finally create (if necessary) and enable the timer.
+  if (!max_stream_duration_timer_) {
+    max_stream_duration_timer_ =
```

Comment:
Just so I understand: where is this being created outside of this code? Do we create one early on in the HCM and this PR is just so that we can adjust the timer as the routing decision changes?